src/helpers.coffee

```diff
@@ -111,12 +111,35 @@ buildLocationData = (first, last) ->
 # This returns a function which takes an object as a parameter, and if that
 # object is an AST node, updates that object's locationData.
 # The object is returned either way.
-exports.addLocationDataFn = (first, last) ->
+exports.addDataToNode = (parserState, first, last) ->
   (obj) ->
-    if ((typeof obj) is 'object') and (!!obj['updateLocationDataIfMissing'])
+    # Add location data
+    if obj?.updateLocationDataIfMissing? and first?
       obj.updateLocationDataIfMissing buildLocationData(first, last)
 
-    return obj
+    # Add comments data
+    unless parserState.tokenComments
+      parserState.tokenComments = {}
+      for token in parserState.parser.tokens when token.comments
+        tokenHash = "#{token[2].first_line}x#{token[2].first_column}-#{token[2].last_line}x#{token[2].last_column}"
+        unless parserState.tokenComments[tokenHash]?
+          parserState.tokenComments[tokenHash] = token.comments
+        else
+          parserState.tokenComments[tokenHash] = parserState.tokenComments[tokenHash].concat token.comments
+
+    if obj.locationData?
+      objHash = "#{obj.locationData.first_line}x#{obj.locationData.first_column}-#{obj.locationData.last_line}x#{obj.locationData.last_column}"
+      if parserState.tokenComments[objHash]?
+        attachCommentsToNode parserState.tokenComments[objHash], obj
+
+    obj
+
+exports.attachCommentsToNode = attachCommentsToNode = (comments, node) ->
+  return unless comments?.length isnt 0
+  node.comments = if node.comments
```

Comment:
Nit: might be clearer as a conditional initialisation + update:

```coffee
node.comments ?= []
node.comments.push comments...
```

I don't feel especially strongly about this, but I think it's a bit easier to digest than a whole multiline `if` statement.